File: leetcode/0273-integer-to-english-words/solution.cpp

const vector<string> thousands({"Thousand", "Million", "Billion"}); // ... more
const vector<string> tens({"Twenty", "Thirty", "Forty", "Fifty", "Sixty",
                           "Seventy", "Eighty", "Ninety"});
const vector<string> smallnumbers({"One", "Two", "Three", "Four", "Five", "Six",
                                   "Seven", "Eight", "Nine", "Ten", "Eleven",
                                   "Twelve", "Thirteen", "Fourteen", "Fifteen",
                                   "Sixteen", "Seventeen", "Eighteen",
                                   "Nineteen"});
const string hundred = "Hundred";
const string zero = "Zero";

class Solution {
public:
    vector<string> thousandsToWords(int num) {
        vector<string> words;
        assert(num >= 0 && num < 1000);
        int hundreds = num / 100;
        if (hundreds > 0) {
            words.push_back(smallnumbers[hundreds - 1]);
            words.push_back(hundred);
        }
        num %= 100;
        if (num > 0 && num <= 19) {
            words.push_back(smallnumbers[num - 1]);
        } else if (num >= 20) {
            words.push_back(tens[num / 10 - 2]);
            num %= 10;
            if (num > 0) {
                words.push_back(smallnumbers[num - 1]);
            }
        }
        return words;
    }
    void extend(vector<string>& v, vector<string> const& v_prime) {
        v.reserve(v.size() + distance(v_prime.begin(), v_prime.end()));
        v.insert(v.end(), v_prime.begin(), v_prime.end());
    }
    string numberToWords(int num) {
        if (num == 0) {
            return zero;
        }
        vector<string> words;
        for (int tho = 0; num > 0; ++tho) {
            auto w = thousandsToWords(num % 1000);
            if (num % 1000 != 0 && tho > 0) {
                w.push_back(tho <= thousands.size() ? thousands[tho - 1]
                                                   : "10^" + to_string(tho * 3));
            }
            extend(w, words);
            words = w;
            num /= 1000;
        }

        return std::accumulate(
            std::next(words.begin()), 
            words.end(), 
            words[0], 
            [](std::string a, std::string b) {
                return a + " " + b;
            }
        );
    }
};

File: leetcode/0273-integer-to-english-words/solution.py
class Solution:
    THOUSANDS = ["Thousand", "Million", "Billion"] # ... more
    TENS = ["Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
    SMALL = ["One", "Two", "Three", "Four", "Five", "Six",
                                   "Seven", "Eight", "Nine", "Ten", "Eleven",
                                   "Twelve", "Thirteen", "Fourteen", "Fifteen",
                                   "Sixteen", "Seventeen", "Eighteen",
                                   "Nineteen"]
    HUNDRED = "Hundred"
    ZERO = "Zero"

    def text_thousand(num: int) -> Optional[str]:
        assert(num >= 0 and num < 1000)
        if num == 0:
            return None
        txt = []
        if num >= 100:
            txt.append(Solution.SMALL[num // 100 - 1] + " Hundred")
        num %= 100
        if num > 0 and num < 20:
            txt.append(Solution.SMALL[num - 1])
        elif num >= 20:
            txt.append(Solution.TENS[num // 10 - 2])
            num %= 10
            if num > 0:
                txt.append(Solution.SMALL[num - 1])
        return " ".join(txt)

    def numberToWords(self, num: int) -> str:
        if num == 0:
            return Solution.ZERO
        thousands = []
        thousands_count = 0
        while num > 0:
            str_thousand = Solution.text_thousand(num % 1000)
            num //= 1000
            if str_thousand:
                if thousands_count > 0:
                    str_thousand += " " + Solution.THOUSANDS[thousands_count - 1]
                thousands.append(str_thousand)
            thousands_count += 1
        return " ".join(reversed(thousands))


        
